### PR TITLE
Add support for page totals and JSON-API page meta data

### PR DESCRIPTION
**Pull request by mjvanleeuwen**
For enhancement proposal https://github.com/yahoo/elide/issues/264.

The page meta data will be returned in the JSON-API response if the page size/limit/number/offset query parameters have been specified.

If the query parameter "page[totals]=true" is also specified, the page totals will be determined and included in the returned page meta data.

### CODE REVIEW

**Comment by yahoocla** [comment]
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

**Comment by yahoocla** [comment]
CLA is valid!

File: elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java [hunk 1]
```diff
@@ -80,9 +89,18 @@ public static Pagination parseQueryParams(final MultivaluedMap<String, String> q
                         } catch (ClassCastException e) {
                             throw new InvalidValueException("page values must be integers");
                         }
+                    } else if (queryParamKey.equals(PAGE_TOTAL_KEY)) {
+                        final String value = paramEntry.getValue().get(0);
+                        if (value.equals("true")) {
```
Comment (DeathByTape) [inline elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java]: Is there a reason we care about the values here? I think it would probably be ok to check for the existence of this parameter (regardless of its value) to enable this feature-- by default it should probably stay disabled anyway.

**Comment by DeathByTape** [comment]
This looks pretty good to me overall; just a couple of additional comments.

First, we should probably have some tests here. Whether it's unit tested or integration tested doesn't matter as much as long as we have looked at some edge cases and so forth. This is less about verifying it right now, but more about not breaking it in the future with additional changes (and having an automated way to test it). Integration tests are usually done through `ResourceIT.java` which is a bit of a mess right now and needs to be cleaned up, but for features that jump across many objects it typically makes more sense.

Next, we need 2 access control points here; one of which you have already implemented.

  1. **Caller access**. This is the access you have implemented here. The caller should be able to selectively request this data from the server only when it needs it.
  1. **Server access**. The server should also be able to control on a per-object basis whether or not it chooses to expose this information to the caller. In some cases, you could imagine that we would _never_ want to leak this cardinality info for some reason. I suggest we implement this through an additional annotation (say `@CanSelectCount` or something?).

This does bring another question @aklish, @wcekan, @clayreimann: if we're going to have a server-side annotation scheme, should we do it for the whole pagination feature? We could start the implementation here (something like `@Paginate(count = true)`) and then expand for other pagination-like things later in additional PR's (to selectively enable/disable pagination for an object, for instance). I am not sure how useful this would actually be in practice, but an `@Paginate` annotation may provide us with a couple of useful things. The most useful feature I could think of for an `@Paginate` (other than enabling/disabling `count`) is for providing **default** limits and other access controls over how pagination is exposed. For example, `@Paginate(defaultLimit = 100, maxLimit = 200)` could be an annotation describing that if no pagination is specified by the caller, the server will assume that is equivalent to paginating to `100` rows and the `maxLimit` would denote that the greatest number of rows for this object we're willing to return at any one time is `200`.

Thoughts?

Thanks for working on this @mjvanleeuwen!

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 2]
```diff
@@ -248,6 +249,20 @@ public void commit() {
...
+    @Override
+    public <T> Long getTotalRecords(Class<T> entityClass, FilterScope filterScope) {
+        final Criterion criterion = filterScope.getCriterion(NOT, AND, OR);
+        final CriteriaExplorer criteriaExplorer = new CriteriaExplorer(
+                entityClass, filterScope.getRequestScope(), criterion);
+        final Criteria sessionCriteria = session.createCriteria(entityClass);
+
+        criteriaExplorer.buildCriteria(sessionCriteria, session);
+
+        sessionCriteria.setProjection(Projections.rowCount());
```
Comment (DeathByTape) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: I think this is equivalent to a `select count(*) ...`, right? So if there was a filter applied, you could select the right set of objects, however, if (by permission) any of those records were filtered out, we would still count them here. I believe the correct value would be something like:

```
this_record_count - total_num_records_filtered
```

This way you're not getting a count for the records which are validly returned by the SQL query, but disallowed by your permissions in Elide.

**Comment by clayreimann** [comment]
I think we _do_ need the `@Paginate` annotation. For the purposes of this PR it should probably just be `@Paginate(countable = true)` but I agree that we should expand it to include at least `@Paginate(defaultLimit = 100, maxLimit = 200)`

File: elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java (hunk 1 above)
Comment (mjvanleeuwen) [inline elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java]: It would seem a bit odd if "page[totals]" value of "false" or "no" was passed and the page totals were still generated.

Also, I am thinking that the client will specify "true" only for the first page request, then with subsequent page requests the totals will normally not need to be recalculated. There is some convenience with being able to just change the value to "false" to stop the totals recalculation.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java (hunk 2 above)
Comment (clayreimann) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: This is a thorny problem, because `total_num_records_filtered` depends on evaluating permissions on EVERY record in SQL. This sounds like something that should be done OOB after the first query. 

The first query would say that this is page 1 of ?? pages and we would kick off a background job that fetches all of the objects from the database. After we evaluate permissions on all of the returned objects and stores the count into a map keyed on user and object requested so that on subsequent requests we can say this is page 2 of 8.

File: elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java [hunk 3]
```diff
@@ -19,30 +19,33 @@
...
+    // For requesting total pages/records be included in the response
+    public static String PAGE_TOTALS_KEY = "page[totals]";
 
     public static final Map<String, PaginationKey> PAGE_KEYS = new HashMap<>();
     static {
         PAGE_KEYS.put("page[size]", PaginationKey.limit);
         PAGE_KEYS.put("page[limit]", PaginationKey.limit);
         PAGE_KEYS.put("page[number]", PaginationKey.page);
         PAGE_KEYS.put("page[offset]", PaginationKey.offset);
+        PAGE_KEYS.put(PAGE_TOTALS_KEY, PaginationKey.totals);
```
Comment (clayreimann) [inline elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java]: While you're here (and since it's a good idea) would you mind replacing the other pagination keys with constants?

File: elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java [hunk 4]
```diff
@@ -83,46 +91,57 @@ public static Pagination parseQueryParams(final MultivaluedMap<String, String> q
...
+    public Pagination evaluate(final Class entityClass) {
+        Paginate paginate = entityClass != null ? (Paginate) entityClass.getAnnotation(Paginate.class) : null;
 
-        boolean isPageBased = pageData.containsKey(PaginationKey.page);
-        int limit = !pageData.containsKey(PaginationKey.limit) ? DEFAULT_PAGE_SIZE : pageData.get(PaginationKey.limit);
-        if (limit < 0) {
-            throw new InvalidValueException("page[size] and page[limit] must contain positive values.");
-        }
+        int defaultLimit = paginate != null ? paginate.defaultLimit() : this.DEFAULT_PAGE_SIZE;
+        int maxLimit = paginate != null ? paginate.maxLimit() : this.MAX_PAGE_SIZE;
```
Comment (clayreimann) [inline elide-core/src/main/java/com/yahoo/elide/core/pagination/Pagination.java]: I'd drop `this` and use the naked variable name, or prefix with the class name if you prefer, since this is a class static.

**Comment by DeathByTape** [comment]
These changes are looking good to me! :) Would you mind squashing at least the `merge upstream changes` commit?

**Comment by DeathByTape** [comment]
The squash is almost perfect, but you merged back into this branch. I would recommend replacing this branch with just those three commits. One easy way to do this is:

```
$ git checkout YOUR_SQUASHED_BRANCH_WITH_ONLY_THOSE_THREE_COMMITS
$ git branch -D add_page_totals_and_meta_page # This will delete your local branch
$ git checkout -b add_page_totals_and_meta_page # This will create a new branch from the one you're currently on and call it 'add_page_totals_and_meta_page'
$ git push <YOUR_REMOTE> add_page_totals_and_meta_page --force
```

This should replace this branch and this PR will only have those 3 commits! :)

Once that is done, I don't see any blocks. @clayreimann does this looks good to you too?

**Comment by DeathByTape** [comment]
This looks good, it just needs to be rebased to resolve conflicts for master. Then I think we can merge this in! 👍

**Comment by DeathByTape** [comment]
:+1: thanks for working on this!
